storage/namespace.go

```diff
@@ -0,0 +1,270 @@
+	if err != nil {
+		return nil, nil, xerrors.NewInvalidParamsError(err)
+	}
+	res, nextPageToken := shard.FetchBlocksMetadata(ctx, limit, pageToken, includeSizes)
+	return res, nextPageToken, nil
+}
+
+func (n *dbNamespace) Bootstrap(
+	bs bootstrap.Bootstrap,
+	writeStart time.Time,
+	cutover time.Time,
+) error {
+	n.Lock()
+	if n.bs == bootstrapped {
+		n.Unlock()
+		return nil
+	}
+	if n.bs == bootstrapping {
+		n.Unlock()
+		return errNamespaceIsBootstrapping
+	}
+	n.bs = bootstrapping
+	n.Unlock()
+
+	if !n.nopts.GetNeedsBootstrap() {
+		n.Lock()
+		n.bs = bootstrapped
+		n.Unlock()
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		err := shard.Bootstrap(bs, n.name, writeStart, cutover)
+		multiErr = multiErr.Add(err)
+	}
+
+	n.Lock()
+	n.bs = bootstrapped
+	n.Unlock()
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) Flush(
+	ctx context.Context,
+	blockStart time.Time,
+	pm persist.Manager,
+) error {
+	n.RLock()
+	if n.bs != bootstrapped {
+		n.RUnlock()
+		return errNamespaceNotBootstrapped
+	}
+	n.RUnlock()
+
+	if !n.nopts.GetNeedsFlush() {
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		// NB(xichen): we still want to proceed if a shard fails to flush its data.
+		// Probably want to emit a counter here, but for now just log it.
+		if err := shard.Flush(ctx, n.name, blockStart, pm); err != nil {
+			detailedErr := fmt.Errorf("shard %d failed to flush data: %v", shard.ID(), err)
+			multiErr = multiErr.Add(detailedErr)
+		}
+	}
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) CleanupFileset(earliestToRetain time.Time) error {
+	if !n.nopts.GetNeedsFilesetCleanup() {
+		return nil
+	}
+
+	multiErr := xerrors.NewMultiError()
+	shards := n.getOwnedShards()
+	for _, shard := range shards {
+		if err := shard.CleanupFileset(n.name, earliestToRetain); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+
+	return multiErr.FinalError()
+}
+
+func (n *dbNamespace) Truncate() {
+	n.initShards()
+
+	// NB(xichen): possibly also clean up disk files and force a GC here to reclaim memory immediately
+}
+
+func (n *dbNamespace) getOwnedShards() []databaseShard {
+	n.RLock()
+	shards := n.shardSet.Shards()
+	databaseShards := make([]databaseShard, len(shards))
+	for i, shard := range shards {
+		databaseShards[i] = n.shards[shard]
+	}
+	n.RUnlock()
+	return databaseShards
+}
+
+func (n *dbNamespace) shardAt(shardID uint32) (databaseShard, error) {
+	n.RLock()
+	shard := n.shards[shardID]
+	n.RUnlock()
+	if shard == nil {
+		return nil, fmt.Errorf("not responsible for shard %d", shardID)
+	}
+	return shard, nil
+}
+
+func (n *dbNamespace) initShards() {
+	shards := n.shardSet.Shards()
+	dbShards := make([]databaseShard, n.shardSet.Max()+1)
+	for _, shard := range shards {
+		dbShards[shard] = newDatabaseShard(shard, n.increasingIndex, n.writeCommitLogFn, n.dopts)
```

Comment:
We should probably pass a no-op function here if `!n.nopts.GetNeedsFlush()` to avoid writes to the commit log for this namespace too.